sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects.scala

```diff
@@ -359,10 +359,27 @@ case class LambdaVariable(value: String, isNull: String, dataType: DataType) ext
 case class MapObjects(
     function: AttributeReference => Expression,
     inputData: Expression,
-    elementType: DataType) extends Expression {
+    elementType: DataType)(
+    completeFunctionCopy: Option[Expression] = None) extends Expression {
 
-  private lazy val loopAttribute = AttributeReference("loopVar", elementType)()
-  private lazy val completeFunction = function(loopAttribute)
+  lazy val completeFunction = completeFunctionCopy.getOrElse {
+    function(AttributeReference("loopVar", elementType)())
+  }
+
+  lazy val loopAttribute = completeFunction.collectFirst {
```

Comment:
here is the tricky part:

* For `TreeNode.transform`, we will go through the `productIterator` to match the children, because we need to make a new copy if transformation changed something.
* For `TreeNode.collect`, we will go through the children directly, so here we may collect the `loopVar` of nested `MapObjects` in `completeFunction`, which is wrong.